libbeat/tests/system/template/template.go

```diff
@@ -0,0 +1,85 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package template
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/elastic/beats/libbeat/asset"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/template"
+	"github.com/elastic/beats/libbeat/version"
+)
+
+// MaxDefaultFieldLength is the limit on the number of default_field values
+// allowed by the test.
+const MaxDefaultFieldLength = 1000
+
+// TestTemplate executes tests on the Beat's index template.
+func TestTemplate(t *testing.T, beatName string) {
+	t.Run("default_field length", testTemplateDefaultFieldLength(beatName))
+}
+
+// testTemplateDefaultFieldLength constructs a template based on the embedded
+// fields.yml data verifies that the length is less than 1000.
+func testTemplateDefaultFieldLength(beatName string) func(*testing.T) {
+	return func(t *testing.T) {
+		// 7.0 is when default_field was introduced.
+		esVersion, err := common.NewVersion("7.0.0")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// Generate a template based on the embedded fields.yml data.
+		tmpl, err := template.New(version.GetDefaultVersion(), beatName, *esVersion, template.TemplateConfig{}, false)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		fieldsBytes, err := asset.GetFields(beatName)
+		if err != nil {
+			t.Fatal("Failed to get embedded fields.yml asset data:", err)
+		}
+
+		fields, err := tmpl.LoadBytes(fieldsBytes)
+		if err != nil {
+			t.Fatal("Failed to load template bytes:", err)
+		}
+
+		templateMap := tmpl.Generate(fields, nil)
+
+		v, _ := templateMap.GetValue("settings.index.query.default_field")
+		defaultValue, ok := v.([]string)
+		if !ok {
+			t.Fatalf("settings.index.query.default_field value has an unexpected type: %T", v)
+		}
+
```

Comment:
What about also executing a elasticsearch query here in addition to catch potential other errors. Say for example ES decides to lower it to 512?

Reply:
That would need to be added somewhere else IMO. I put the test here so that it could be executed quickly as a unit test and not have an external dependency on ES.

This could be put into system test. Each beat should probably call setup to install its template and ILM policy, index a document, then perform a query.

Reply:
Agree, this would fit well here: https://github.com/elastic/beats/blob/master/filebeat/tests/system/test_base.py#L30 Perhaps someone in the team from @urso could follow up with this?

Reply:
@andrewkroh @ruflin I've created https://github.com/elastic/beats/issues/14352 to make a change in libbeat to be more proactive to detect that error.